File: IsometricRPG/IsometricRPG/Utils/PerlinNoise.swift
import CoreGraphics

/// 2D Perlin-style noise generator for procedural terrain.
/// Uses a gradient-based approach with smoothstep interpolation.
struct PerlinNoise {
    private let permutation: [Int]
    private let gradients: [(CGFloat, CGFloat)]

    init(seed: UInt64) {
        var rng = SeededRNG(seed: seed)

        // Build a shuffled permutation table (0..<256, doubled for wrapping)
        var perm = Array(0..<256)
        for i in stride(from: 255, through: 1, by: -1) {
            let j = Int(rng.next() % UInt64(i + 1))
            perm.swapAt(i, j)
        }
        permutation = perm + perm

        // Precompute unit-circle gradient vectors
        var grads: [(CGFloat, CGFloat)] = []
        for _ in 0..<256 {
            let angle = CGFloat(rng.next() % 3600) / 3600.0 * .pi * 2
            grads.append((cos(angle), sin(angle)))
        }
        gradients = grads
    }

    /// Sample noise at a point. Returns a value roughly in -1...1.
    func sample(x: CGFloat, y: CGFloat) -> CGFloat {
        let xi = Int(floor(x)) & 255
        let yi = Int(floor(y)) & 255
        let xf = x - floor(x)
        let yf = y - floor(y)

        let u = fade(xf)
        let v = fade(yf)

        let aa = permutation[permutation[xi] + yi]
        let ab = permutation[permutation[xi] + yi + 1]
        let ba = permutation[permutation[xi + 1] + yi]
        let bb = permutation[permutation[xi + 1] + yi + 1]

        let x1 = lerp(grad(hash: aa, x: xf, y: yf),
                       grad(hash: ba, x: xf - 1, y: yf), t: u)
        let x2 = lerp(grad(hash: ab, x: xf, y: yf - 1),
                       grad(hash: bb, x: xf - 1, y: yf - 1), t: u)

        return lerp(x1, x2, t: v)
    }

    /// Multi-octave fractal noise for more natural terrain.
    func fractal(x: CGFloat, y: CGFloat, octaves: Int = 4,
                 persistence: CGFloat = 0.5, lacunarity: CGFloat = 2.0) -> CGFloat {
        var total: CGFloat = 0
        var amplitude: CGFloat = 1
        var frequency: CGFloat = 1
        var maxAmplitude: CGFloat = 0

        for _ in 0..<octaves {
            total += sample(x: x * frequency, y: y * frequency) * amplitude
            maxAmplitude += amplitude
            amplitude *= persistence
            frequency *= lacunarity
        }

        return total / maxAmplitude
    }

    private func fade(_ t: CGFloat) -> CGFloat {
        return t * t * t * (t * (t * 6 - 15) + 10)
    }

    private func lerp(_ a: CGFloat, _ b: CGFloat, t: CGFloat) -> CGFloat {
        return a + t * (b - a)
    }

    private func grad(hash: Int, x: CGFloat, y: CGFloat) -> CGFloat {
        let idx = hash & 255
        let (gx, gy) = gradients[idx]
        return gx * x + gy * y
    }
}

File: IsometricRPGMono/Utils/PerlinNoise.cs
namespace IsometricRPG;

public class PerlinNoise
{
    private readonly int[] _permutation;
    private readonly (float gx, float gy)[] _gradients;

    public PerlinNoise(ulong seed)
    {
        var rng = new SeededRNG(seed);

        // Build permutation table
        var perm = Enumerable.Range(0, 256).ToArray();
        for (int i = 255; i >= 1; i--)
        {
            int j = (int)(rng.Next() % (ulong)(i + 1));
            (perm[i], perm[j]) = (perm[j], perm[i]);
        }
        // Double the table to avoid boundary checks
        _permutation = new int[512];
        for (int i = 0; i < 512; i++)
            _permutation[i] = perm[i & 255];

        // Build gradient table
        _gradients = new (float, float)[256];
        for (int i = 0; i < 256; i++)
        {
            float angle = (float)(rng.Next() % 3600UL) / 3600.0f * MathF.PI * 2f;
            _gradients[i] = (MathF.Cos(angle), MathF.Sin(angle));
        }
    }

    public float Sample(float x, float y)
    {
        int xi = (int)MathF.Floor(x) & 255;
        int yi = (int)MathF.Floor(y) & 255;

        float xf = x - MathF.Floor(x);
        float yf = y - MathF.Floor(y);

        float u = Fade(xf);
        float v = Fade(yf);

        int aa = _permutation[_permutation[xi]     + yi];
        int ab = _permutation[_permutation[xi]     + yi + 1];
        int ba = _permutation[_permutation[xi + 1] + yi];
        int bb = _permutation[_permutation[xi + 1] + yi + 1];

        float x1 = Lerp(Grad(aa, xf,       yf),       Grad(ba, xf - 1f, yf),       u);
        float x2 = Lerp(Grad(ab, xf,       yf - 1f), Grad(bb, xf - 1f, yf - 1f), u);

        return Lerp(x1, x2, v);
    }

    public float Fractal(float x, float y, int octaves = 4, float persistence = 0.5f, float lacunarity = 2.0f)
    {
        float total = 0f;
        float frequency = 1f;
        float amplitude = 1f;
        float maxValue = 0f;

        for (int i = 0; i < octaves; i++)
        {
            total += Sample(x * frequency, y * frequency) * amplitude;
            maxValue += amplitude;
            amplitude *= persistence;
            frequency *= lacunarity;
        }

        return total / maxValue;
    }

    private float Fade(float t) => t * t * t * (t * (t * 6f - 15f) + 10f);

    private float Lerp(float a, float b, float t) => a + t * (b - a);

    private float Grad(int hash, float x, float y)
    {
        int idx = hash & 255;
        var (gx, gy) = _gradients[idx];
        return gx * x + gy * y;
    }
}
